src/matrix/room/timeline/persistence/MemberWriter.js

```diff
@@ -0,0 +1,203 @@
+/*
+Copyright 2021 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import {MemberChange, RoomMember, EVENT_TYPE as MEMBER_EVENT_TYPE} from "../../members/RoomMember.js";
+import {LRUCache} from "../../../../utils/LRUCache.js";
+
+export class MemberWriter {
+    constructor(roomId) {
+        this._roomId = roomId;
+        this._cache = new LRUCache(5, member => member.userId);
+    }
+
+    writeTimelineMemberEvent(event, txn) {
+        return this._writeMemberEvent(event, false, txn);
+    }
+
+    writeStateMemberEvent(event, isLimited, txn) {
+        // member events in the state section when the room response
+        // is not limited must always be lazy loaded members.
+        // If they are not, they will be repeated in the timeline anyway.
+        return this._writeMemberEvent(event, !isLimited, txn);
+    }
+
+    async _writeMemberEvent(event, isLazyLoadingMember, txn) {
+        const userId = event.state_key;
+        if (!userId) {
+            return;
+        }
+        const member = RoomMember.fromMemberEvent(this._roomId, event);
+        if (!member) {
+            return;
+        }
+
+        let existingMember = this._cache.get(userId);
+        if (!existingMember) {
+            const memberData = await txn.roomMembers.get(this._roomId, userId);
+            if (memberData) {
+                existingMember = new RoomMember(memberData);
+            }
+        }
+
+        // either never heard of the member, or something changed
+        if (!existingMember || !existingMember.equals(member)) {
```

Comment:
if there is a display name change, and both state and timeline contain this event (which synapse does), and the room response is not limited, this will fail to emit a member change when the timeline event is processed because the state event already updated the cache/storage so the equals check passes ...